python/tank_vendor/shotgun_deploy/import_handler.py

```diff
@@ -0,0 +1,283 @@
+        >>> sys.meta_path.append(importer)
+        >>> # import/run a bunch of code
+        >>> # context change, need to use a different core
+        >>> importer.set_core_path("/path/to/a/different/version/of/tk-core")
+        >>> # namespaces cleared. new imports will come from new core location
+
+    When an instance of this object is added to `sys.meta_path`, it is used to
+    alter the way python imports packages.
+
+    The core path is used to locate modules attempting to be loaded. The core
+    path can be set via `set_core_path` to alter the location of existing and
+    future core imports.
+
+    For more information on custom import hooks, see PEP 302:
+        https://www.python.org/dev/peps/pep-0302/
+
+    """
+
+    def __init__(self, core_path):
+        """Initialize the custom importer.
+
+        :param core_path: A str path to the core location to import from.
+
+        """
+
+        self._core_path = None
+        self._namespaces = []
+
+        # a dictionary to hold module information after it is found, before
+        # it is loaded.
+        self._module_info = {}
+
+        # re-imports any existing modules for the core namespaces
+        self.set_core_path(core_path)
+
+    def __repr__(self):
+        """
+        A unique representation of the handler.
+
+        :return: str representation.
+        """
+
+        return (
+            "<CoreImportHandler for core located in: '%s'" % (self.core_path,))
+
+    def find_module(self, module_fullname, package_path=None):
+        """Locates the given module in the current core.
+
+        :param module_fullname: The fullname of the module to import
+        :param package_path: None for a top-level module, or
+            package.__path__ for submodules or subpackages
+
+        The package_path is currently ignored by this method as it ensures we're
+        importing the module from the current core path.
+
+        For further info, see the docs on find_module here:
+            https://docs.python.org/2/library/imp.html#imp.find_module
+
+        :returns: this object (also a loader) if module found, None otherwise.
+        """
+
+        # get the package name (first part of the module fullname)
+        module_path_parts = module_fullname.split('.')
+        package_name = module_path_parts[0]
+
+        # make sure the package is in the list of namespaces before continuing.
+        if package_name not in self._namespaces:
+            # the package is not in one of the core namespaces. returning
+            # None tells python to use the next importer available (likely the
+            # default import mechanism).
+            return None
+
+        # module path without the target module name
+        module_name = module_path_parts.pop()
+
+        # partial path to the module on disk
+        if len(module_path_parts):
+            module_path = os.path.join(*module_path_parts)
```

Comment:
I don't think it matters at this point, but would this break with any "import foo as bar" style imports?
Is there a way to get __file__ from the module to better compute the path?